ws/api/endpoints.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/go-zoo/bone"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux/internal/apiutil"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/ws"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			encodeError(&req, w, err)
+			return
+		}
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+		req.conn = conn
+		client := ws.NewClient(conn, "")
+
+		if err := svc.Subscribe(context.Background(), req.thingKey, req.chanID, req.subtopic, client); err != nil {
+			fmt.Println("Error in subscribe")
+			logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+			encodeError(&req, w, err)
+		}
+
+		// In case previous subscription wasn't closed properly
+		if conn == nil {
+			conn, err = upgrader.Upgrade(w, r, nil)
+			if err != nil {
+				logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+				return
+			}
+			client = ws.NewClient(conn, client.GetID())
+		}
+
+		logger.Debug(fmt.Sprintf("Successfully upgraded communication to WS on channel %s", req.chanID))
+		msgs := make(chan []byte)
+
+		// Listen for messages received from the chan messages, and publish them to broker
+		go client.Publish(svc, logger, req.thingKey, req.chanID, req.subtopic, msgs)
```

Comment:
Since the adapter is already using Client in the Subscribe, it would be nice to decouple them here (so that client does not use the adapter in Publish. You can do something similar to `listen` (even at the cost of exposing messaging package to the API layer).